### PR TITLE
Add support for logging 

### LINKED ISSUE #30: Add support for the filebeat logstash interface or any other log collection mechanism that would work with machine charms
### Enhancement Proposal

In the LMA stack, a combination of Filebeat, Graylog and Elasticsearch charms can be used to aggregate and store logs.

COS Proxy implements relations for Grafana dashboards, Prometheus scrape targets, there's also a solution to get NRPE checks to work, but there's no 1:1 replacement for the filebeat's logstash relation. This is a potential blocked for migrating from the LMA stack to COS.

### PR DESCRIPTION
**Pull request by rbarry82**
## Issue
`cos-proxy` did not previously have any support for logging in reactive/machine charms over the `filebeat` interface.

In addition, the output from NRPE commands was not visible, so operators had to manually check on failed jobs.

## Solution
Use [vector](https://vector.dev/) to provide a filebeat-compatible ingestor.

Add support for remote Loki relations as a "sink", and manipulate the input from filebeat and nrpe-exporter journald logs using vector remap language before pushing it off to Loki.

Vector is able to use a lookaside "enrichment table" to provide ancillary data. Since the NRPE exporter itself is not aware of (nor has reason to be) juju topology, use the ip address and command concatenated as a composite key for enrichment table lookups.

Closes #30 

## Testing Instructions
Deploy a machine charm. Deploy and relate to a filebeat subordinate and/or an NRPE. Watch the systemd journal after `juju ssh`-ing into the cos-proxy charm, or relate to Loki and check there.

## Release Notes
Add support for logging

### CODE REVIEW

File: charmcraft.yaml [hunk 1]
```diff
@@ -17,3 +17,19 @@ parts:
...
+    override-pull: |
+      VERSION=0.27.0
+      if [ $CRAFT_TARGET_ARCH == "amd64" ]; then
+        ARCH=x86_64-unknown-linux-musl
+      elif [ $CRAFT_TARGET_ARCH == "aarch64" ]; then
+        ARCH=aarch64-unknown-linux-musl
+      fi
+
+      VECTOR_URI="https://packages.timber.io/vector/${VERSION}/vector-${VERSION}-${ARCH}.tar.gz"
+      curl -sSfL ${VECTOR_URI} | tar xzf - -C . --strip-components=3 ./vector-${ARCH}/bin/vector
```
Comment (sed-i) [inline charmcraft.yaml:35]: Should we be concerned here about (in)visibility to CVE scanning?

File: lib/charms/vector/v0/vector.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+
+class VectorEvents(ObjectEvents):
+    """Event descriptor for events raised by `NrpeExporterProvider`."""
+
+    config_changed = EventSource(VectorConfigChangedEvent)
+
+
+class VectorProvider(Object):
+    """An provider for Vector, an observability swiss-army knife."""
```
Comment (sed-i) [inline lib/charms/vector/v0/vector.py:183]:
```suggestion
    """A provider for Vector, an observability swiss-army knife."""
```

File: src/charm.py [hunk 3]
```diff
@@ -198,6 +208,118 @@ def _nrpe_relation_broken(self, _):
         self._stored.have_nrpe = False
         self._set_status()
 
+    def _on_filebeat_relation_joined(self, event):
+        self._stored.have_filebeat = True
+
+        # Make sure the vector binary is present with a systemd service
+        if not Path("/usr/local/bin/vector").exists():
```
Comment (sed-i) [inline src/charm.py:215]: Should this entire block (apart from systemd calls) go under `install` or charm lib?

Comment (rbarry82) [reply src/charm.py:215]: No, manipulating files on disk shouldn't be in the charm lib. As much as it's not actually handling relation data in this case.

We can't _really_ minimize resource utilization (it's not a subordinate), but like the exporter, we're treating this almost like socket-based activation and starting it as needed.

File: charmcraft.yaml (hunk 1 above)
Comment (rbarry82) [reply charmcraft.yaml:35]: Considering the very narrow scope of what it actually does and that it's from DataDog, sort-of yes, but there's not a latest release.

**Pull request merged**